### PR TITLE
Purge cloudflare cache for collective pages on certain actions

### PR DESCRIPTION
**Pull request by Betree**
Resolve https://github.com/opencollective/opencollective/issues/1870

### CODE REVIEW

File: server/gateways/cloudflare.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import cloudflare from 'cloudflare';
+import { get } from 'lodash';
+
+import logger from '../lib/logger';
+
+// Load config
+const cfConfig = {
+  email: get(config, 'clouflare.email'),
+  key: get(config, 'clouflare.key'),
+};
```
Comment (znarf) [inline server/gateways/cloudflare.js]: Looks unnecessary compared to `config.cloudflare`.

File: server/gateways/cloudflare.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+export const zone = get(config, 'clouflare.zone');
```
Comment (znarf) [inline server/gateways/cloudflare.js]: Why should that be exported?

File: server/gateways/cloudflare.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+// Check config
+const isLiveServer = ['staging', 'production'].includes(config.env);
+const hasConfig = Boolean(cfConfig.email && cfConfig.key && zone);
+
+if (isLiveServer && !hasConfig) {
+  logger.warn('Your Cloudflare config is imcomplete, you must provide an email, a key and a zone');
```
Comment (znarf) [inline server/gateways/cloudflare.js]: Let's avoid "you" for such messages?

File: server/gateways/cloudflare.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import config from 'config';
```
Comment (znarf) [inline server/gateways/cloudflare.js]: Why not put this library in `lib` instead of `gateways`?

**Comment by znarf** [comment]
Looks good, let's try on staging before merging.

File: server/gateways/cloudflare.js (hunk 2 above)
Comment (Betree) [reply server/gateways/cloudflare.js]: Almost all queries from `cloudflare` library expect a `zone` to be passed in params

File: server/gateways/cloudflare.js (hunk 1 above)
Comment (Betree) [reply server/gateways/cloudflare.js]: This object is to get passed to `cloudflare(...)` (the lib) and it only expect an `email` and a `key`. If we pass the full object it will also include the `zone` parameter.

However I understand that this can be confusing, I'll refactor

File: server/gateways/cloudflare.js (hunk 2 above)
Comment (Betree) [reply server/gateways/cloudflare.js]: But I'll remove the cloudflare export to only export the methods that will be used

File: server/gateways/cloudflare.js (hunk 4 above)
Comment (Betree) [reply server/gateways/cloudflare.js]: `gateways` seemed to me like the place where we wrap third-party libraries. What's the purpose of this folder if that's not for that?

Comment (znarf) [reply server/gateways/cloudflare.js]: Then I would drop `gateways` and have everything in `lib`, this is where most wrappers are already.

**Comment by Betree** [comment]
Tests on staging went well. Also added the configuration to prod.

:warning: Cloudflare doesn't allow to generate multiple API KEYs, so the key is shared between staging and production.

File: server/lib/cloudflare.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+// Load config
+const cfConfig = config.cloudflare || {};
+
+// Check config
+const isLiveServer = ['staging', 'production'].includes(config.env);
+const hasConfig = Boolean(cfConfig.email && cfConfig.key && cfConfig.zone);
+
+if (isLiveServer && !hasConfig) {
+  logger.warn('Your Cloudflare config is imcomplete, you must provide an email, a key and a zone');
```
Comment (znarf) [inline server/lib/cloudflare.js]: Same comment than the other day about "your", don't think we should formulate errors/warnings like that.

File: server/lib/cloudflare.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * Purge the given page from cloudflare's cache. In dev environments this function
+ * will only log to the console and will skip the call to cloudflare.
+ *
+ * @param {string|array} pagePaths - a path or an array of paths `/eslint`, `['/', '/about']`
+ * @returns {Promise}
+ */
+export const purgeCacheForPage = pagePaths => {
+  const prepareURL = pagePath => {
+    const prependSlash = pagePath[0] === '/' ? '' : '/';
+    return `${config.host.website}${prependSlash}${pagePath}`;
```
Comment (znarf) [inline server/lib/cloudflare.js]: Have you tested?

My experience of Cloudflare is that we need to empty `https://frontend.opencollective.com/babel` not `https://opencollective.com/babel`. I personally have the habit of purging both, not sure it's justified.

Comment (Betree) [reply server/lib/cloudflare.js]: I tested a bit naively, just ensured the collective page was properly updated.

I just did more advanced testing with `curl -svo /dev/null` to get the headers and it seems like I was lucky during my tests; the cache is not purged properly.

However I don't have more luck by doing:

```es6
api.zones.purgeCache('...', {files: ['https://staging.opencollective.com/captainfact_io', 'https://staging-frontend.opencollective.com/captainfact_io']})
```

Comment (znarf) [reply server/lib/cloudflare.js]: It's `https://frontend-staging.opencollective.com`.
